### ISSUE TITLE
Revise authentication parameters for connector methods

### ISSUE THREAD

**Issue by diego-plan9** [Qiskit/qiskit-ibmq-provider#34]
### What is the expected behavior?

Currently, most of the methods of `IBMQConnector` that interact with and endpoint accept the credentials information (`hub`, `group`, `user_id`, etc) as parameters, and perform a similar set of operations at the beginning of the function:

```python
        if access_token:
            self.req.credential.set_token(access_token)
        if user_id:
            self.req.credential.set_user_id(user_id)
        if not self.check_credentials():
            raise CredentialsError('credentials invalid')
```

However, the credentials information is already present in `self.config`, which makes the parameters and some checks redundant (and actually, the parameters are unused by the rest of the codebase). It seems a good candidate for simplification, by just removing the parameters and fetching the needed information from `self.config` directly.

**Comment by jwoehr** [comment]
It _does_ look like those parameters are only used within the package itself.

**Comment by diego-plan9** [comment]
Yes - the issue is mostly about reorganizing things a bit to reduce complexity, taking advantage of that fact indeed!

**Comment by jwoehr** [comment]
A little confused here!

Qiskit 8.0 has IBMQConnector.get_job() as:
`def get_job_url(config, hub=None, group=None, project=None):`

Current has IBMQConnector.get_job() as:
`def get_job(self, id_job, hub=None, group=None, project=None,
                exclude_fields=None, include_fields=None,
                access_token=None, user_id=None)`

So apparently fields were still being added to the code until recently.

Which of these fields is it desirable to eliminate and always fetch from the config? All of them?

- hub
- group
- project
- access_token
- user_id

??

**Comment by diego-plan9** [comment]
> Which of these fields is it desirable to eliminate and always fetch from the config? All of them?

The ones that can be inferred directly from `self.config` and are only used in the `if` "block" mentioned on the issue description  - actually, the list you posted right after your sentence (`hub`, `group`, `project`, `access_token`, `user_id`) looks about right in the general case!

In the example you mentioned (`IBMQConnector.get_job()`), `id_job`, `exclude_fields` and `include_fields` are parameters that are not related to authentication and are used in order to provide the functionality required by the method - those are good examples of parameters that should _not_ be removed :+1:

**Comment by jwoehr** [comment]
So for example this:
```
def get_job_url(config, hub=None, group=None, project=None):
    """Return the URL for a job."""
    hub = config.get('hub', hub)
    group = config.get('group', group)
    project = config.get('project', project)

    if hub and group and project:
        return '/Network/{}/Groups/{}/Projects/{}/jobs'.format(hub, group,
                                                               project)
    return '/Jobs'
```
would become this:
```
def get_job_url():
    """Return the URL for a job."""
    hub = self.config.get('hub', None)
    group = self.config.get('group', None)
    project = self.config.get('project', None)

    if hub and group and project:
        return '/Network/{}/Groups/{}/Projects/{}/jobs'.format(hub, group,
                                                               project)
    return '/Jobs'
```
... with the `None`s possibly being redundant but harmless. Correct?

**Comment by diego-plan9** [comment]
In essence yes, but that might not be the best example to use  - please note that the issue was originally about most of the methods of the `IBMQConnector` class (as that class is the one containing the `self.config`).

`get_job_url()` is actually a module-level function that does not belong to the class, and though I think it will also be tweaked a bit as a byproduct of this issue, I'd suggest starting with the `IBMQConnector.foo()` methods that contain blocks similar to the one quoted in the issue description:

* `IBMQConnector.run_job()`
* `IBMQConnector.get_job()`
* `IBMQConnector.get_jobs()`
* `IBMQConnector.get_status_job()`
* `IBMQConnector.get_status_jobs()`
* `IBMQConnector.cancel_job()`
* `IBMQConnector.backend_status()`
* `IBMQConnector.backend_properties()`
* `IBMQConnector.available_backends()`

The idea would be to examine them in a case-by-case basis, and remove the superfluous parameters from those:
* `access_token` and `user_id` are probably only used for the `if access_token` and `if user_id` in the block, and hopefully safe to remove directly along with those `if`s (as the callers of the function are not really using them).
* `hub`, `group` and `project` might be used for calling the module-level functions (ie. `get_foo_url()`), but they are also probably safe to remove as well since `config` is also a parameter to those functions.
* in general, the rest of the parameters are not authentication-related and are needed for the function's logic, and as such they would stay.

So, for example, `IBMQConnector.cancel_job` might end up looking like:
```python
    def cancel_job(self, id_job):
        """Cancel the information about a job, by its id."""
        if not self.check_credentials():
            return {"error": "Not credentials valid"}

        url = get_job_url(self.config)

        url += '/{}/cancel'.format(id_job)

        res = self.req.post(url)

        return res
```

I'm including a bunch of "probably", "hopefully" in the comment, as the task has some exploratory side to it to verify these claims, and adjust the rest of the codebase accordingly - but hopefully they are a good guide for the issue! There are other niceties we can apply (for example, instead of repeating the `if not self.check_credentials(): ...` block, use a custom decorator), but just removing the parameters would improve maintainability by quite a bit.

**Comment by jwoehr** [comment]
Are the `qiskit-ibmq-provider/test/ibmq` tests sufficient, or after making changes will I have to get creative?

I suspect I know the answer to this question :grin:

**Comment by diego-plan9** [comment]
Actually I _think_ they will suffice :slightly_smiling_face:

**Comment by jwoehr** [comment]
Thank you. I'll do this.

**Comment by diego-plan9** [comment]
Fixed via #56
